// @semoss/sdk-react/dist/index.js
import { createContext, useMemo, useState, useEffect, useContext } from 'react';
import { jsx } from 'react/jsx-runtime';
import { unstable_batchedUpdates } from 'react-dom';
import { Insight } from '@semoss/sdk';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */


function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

/**
 * Context of the react data
 */
const InsightContext = createContext(undefined);
const InsightProvider = (props) => {
    const { children, options } = props;
    // create the new insight on load
    const insight = useMemo(() => {
        return new Insight();
    }, []);
    const [isInitialized, setIsInitialized] = useState(false);
    const [isAuthorized, setIsAuthorized] = useState(false);
    const [isReady, setIsReady] = useState(false);
    const [error, setError] = useState(null);
    const [system, setSystem] = useState(null);
    const [insightId, setInsightId] = useState('');
    /**
     * Sync the insight with react
     */
    const syncInsight = () => __awaiter(void 0, void 0, void 0, function* () {
        unstable_batchedUpdates(() => {
            setError(insight.error);
            setSystem(insight.system);
            setIsAuthorized(insight.isAuthorized);
            setIsInitialized(insight.isInitialized);
            setIsReady(insight.isReady);
            setInsightId(insight.insightId);
        });
    });
    const wrappedActions = useMemo(() => {
        return Object.keys(insight.actions).reduce((acc, val) => {
            //@ts-expect-error TODO Fix Typing
            acc[val] = (...args) => __awaiter(void 0, void 0, void 0, function* () {
                // wait for the action to complete
                //@ts-expect-error TODO Fix Typing
                const response = yield insight.actions[val].apply(null, [
                    ...args,
                ]);
                // sync it
                syncInsight();
                // return the response
                return response;
            });
            return acc;
        }, {});
    }, [insight, insight.actions]);
    // initialize the insight / destroy
    useEffect(() => {
        // initialize the insight
        insight.initialize(options).finally(() => {
            // update the state
            syncInsight();
        });
        return () => {
            // destroy the insight
            insight.destroy().finally(() => {
                // update the state
                syncInsight();
            });
        };
    }, [insight, options]);
    return (jsx(InsightContext.Provider, Object.assign({ value: {
            isInitialized: isInitialized,
            isAuthorized: isAuthorized,
            isReady: isReady,
            error: error,
            system: system,
            actions: wrappedActions,
            insightId: insightId
        } }, { children: children })));
};

/**
 * Access the current insight
 */
const useInsight = () => {
    const context = useContext(InsightContext);
    if (context === undefined) {
        throw new Error('useInsight must be used within InsightProvider');
    }
    return context;
};

export { InsightProvider, useInsight };
